util: Revert back `MoveFileExW` call for MinGW-w64

hebasto opened the pull request:
Unfortunately, bitcoin/bitcoin#24308 introduced a [regression](https://github.com/bitcoin/bitcoin/pull/24308#issuecomment-1037259386) for mingw builds.

The root of the problem is a broken implementation of [`std::filesystem::rename`](https://en.cppreference.com/w/cpp/filesystem/rename). In particular, the expected behavior
> If `old_p` is a non-directory file, then `new_p` must be ... existing non-directory file: `new_p` _is first deleted_...

fails with the "File exists" error.

This PR reverts back the `MoveFileExW` call, and adds the [suggested](https://github.com/bitcoin/bitcoin/pull/24308#pullrequestreview-878832906) unit test.

hebasto committed:
util: Revert back MoveFileExW call for MinGW-w64

hebasto committed:
test: Add fs_tests/rename unit test

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

fanquake commented:
> The root of the problem is a broken implementation of [std::filesystem::rename](https://en.cppreference.com/w/cpp/filesystem/rename).

What version is broken? Is there an upstream issue / bug report?

MarcoFalke commented:
Ah, it fails: https://cirrus-ci.com/task/6670169794674688?logs=ci#L4536

Concept ACK

laanwj commented:
Good catch… Leave it to windows libraries to not actually implement the standard as it is defined.

MarcoFalke commented:
At one point it might be better to drop support for mingw and use clang instead for cross compiling. With all the bugs we are seeing and other projects moving to clang I am not convinced this is a good long term choice anymore.

laanwj commented on src/util/system.cpp:
```diff
@@ -1062,9 +1062,14 @@ void ArgsManager::LogArgs() const
 bool RenameOver(fs::path src, fs::path dest)
 {
+#ifdef __MINGW64__
+    return MoveFileExW(src.wstring().c_str(), dest.wstring().c_str(),
```
Might want to add a comment here specifying why this workaround exists, and we should probably file an upstream issue and refer to it, so it has a limited lifetime.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24331#issuecomment-1039185324).

laanwj commented:
I didn't realize the bug was Mingw only. So MSVC (or even clang) implements this correctly?

> At one point it might be better to drop support for mingw and use clang instead for cross compiling.

Agree. If it's less buggy than mingw-w64, I think it would be good to consider using clang for the windows build.

vasild commented on src/test/fs_tests.cpp:
```diff
@@ -118,4 +118,37 @@ BOOST_AUTO_TEST_CASE(fsbridge_fstream)
+        std::string contents;
+        file >> contents;
+        BOOST_CHECK_EQUAL(contents, path1_contents);
+    }
```
It is a good practice to clean up after the test, even if it is a temporary folder. I forgot that, but woke up during the night with the thought "file cleanup was forgotten!".

```suggestion
    }
    fs::remove(path2);
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24331#issuecomment-1038872961).

vasild reviewed (approved):
ACK 15a043dcee116f5273a262ad83853c791c50b9fa

vasild commented:
mingw32 works ok? Or we don't support that?

hebasto commented:
> mingw32 works ok? Or we don't support that?

The latter, I [guess](https://github.com/bitcoin/bitcoin/pull/15939).

hebasto commented:
Updated 15a043dcee116f5273a262ad83853c791c50b9fa -> 4d92107041c508cf7bca47dce088979e519fb8bc ([pr24331.01](https://github.com/hebasto/bitcoin/commits/pr24331.01) -> [pr24331.02](https://github.com/hebasto/bitcoin/commits/pr24331.02), [diff](https://github.com/hebasto/bitcoin/compare/pr24331.01..pr24331.02)):

- addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/24331#discussion_r805652748)

vasild reviewed (approved):
ACK 4d92107041c508cf7bca47dce088979e519fb8bc

fanquake commented:
> What version is broken? Is there an upstream issue / bug report?

laanwj commented:
> mingw32 works ok? Or we don't support that?

afaik mingw-w64 was a fork years ago because mingw32 wasn't really maintained anymore, i doubt they support newer C++ standards required

hebasto commented:
Updated 4d92107041c508cf7bca47dce088979e519fb8bc -> dc01cbc538765f64326bca30952c83e3862d0d54 ([pr24331.02](https://github.com/hebasto/bitcoin/commits/pr24331.02) -> [pr24331.03](https://github.com/hebasto/bitcoin/commits/pr24331.03), [diff](https://github.com/hebasto/bitcoin/compare/pr24331.02..pr24331.03)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/24331#discussion_r805632060)

vasild reviewed (approved):
ACK dc01cbc538765f64326bca30952c83e3862d0d54

fanquake commented:
> What version(s) are broken? Is there an upstream issue / bug report?

This is still unclear to me?

laanwj commented:
> This is still unclear to me?

FWIW there's a comment in the source
```
    // This bug has been fixed in upstream:
    //  - GCC 10.3: 8dd1c1085587c9f8a21bb5e588dfe1e8cdbba79e
    //  - GCC 11.1: 1dfd95f0a0ca1d9e6cbc00e6cbfd1fa20a98f312
```
